mixins/property-required/property-required-mixin.js

```diff
@@ -0,0 +1,108 @@
+import { dedupeMixin } from '@open-wc/dedupe-mixin';
+
+const TIMEOUT_DURATION = 3000;
+
+export function createDefaultMessage(tagName, propertyName) {
+	return `${tagName}: "${propertyName}" attribute is required.`;
+}
+
+export function createUndefinedPropertyMessage(tagName, propertyName) {
+	return `PropertyRequiredMixin: "${tagName.toLowerCase()}" has no property "${propertyName}".`;
+}
+
+export function createInvalidPropertyTypeMessage(tagName, propertyName) {
+	return `PropertyRequiredMixin: only String properties can be required ("${tagName}" required property "${propertyName}".`;
+}
+
+export const PropertyRequiredMixin = dedupeMixin(superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._allProperties = new Map();
+		this._requiredProperties = new Map();
+		this._initProperties(Object.getPrototypeOf(this));
+	}
+
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+		for (const name of this._requiredProperties.keys()) {
+			this._validateRequiredProperty(name);
+		}
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+		this._requiredProperties.forEach((value, name) => {
+			const doValidate = changedProperties.has(name) ||
+			value.dependentProps.includes(name);
+			if (doValidate) this._validateRequiredProperty(name);
+		});
+	}
+
+	addRequiredProperty(name, opts) {
+
+		const prop = this._allProperties.get(name);
+		if (prop === undefined) {
+			throw new Error(createUndefinedPropertyMessage(this.tagName.toLowerCase(), name));
+		}
+
+		if (prop.type !== String) {
+			throw new Error(createInvalidPropertyTypeMessage(this.tagName.toLowerCase(), name));
+		}
+
+		opts = {
+			...{ dependentProps: [], message: defaultMessage => defaultMessage, validator: hasValue => hasValue },
+			...opts
+		};
+
+		this._requiredProperties.set(name, {
+			attrName: prop.attribute || name,
+			dependentProps: opts.dependentProps,
+			message: opts.message,
+			thrown: false,
+			timeout: null,
+			validator: opts.validator
+		});
+
+	}
+
+	flushRequiredPropertyErrors() {
```

Comment:
This method is just here for testing, allowing them to bypass the 3s delay and have the exceptions be thrown in their context.